File: runtime/src/main/java/com/jaynewstrom/json/runtime/internal/ListDeserializer.java
package com.jaynewstrom.json.runtime.internal;

import com.fasterxml.jackson.core.JsonParser;
import com.fasterxml.jackson.core.JsonToken;
import com.jaynewstrom.json.runtime.JsonDeserializer;
import com.jaynewstrom.json.runtime.JsonDeserializerFactory;

import java.io.IOException;
import java.util.ArrayList;
import java.util.List;

public final class ListDeserializer<T> implements JsonDeserializer<List<T>> {
    private final JsonDeserializer<T> elementFactory;

    public ListDeserializer(JsonDeserializer<T> elementFactory) {
        this.elementFactory = elementFactory;
    }

    @Override public List<T> deserialize(JsonParser jp, JsonDeserializerFactory deserializerFactory) throws IOException {
        if (jp.currentToken() != JsonToken.START_ARRAY) {
            throw new IOException("Expected data to start with an Array");
        }

        List<T> list = new ArrayList<>();

        while (jp.nextToken() != JsonToken.END_ARRAY) {
            list.add(elementFactory.deserialize(jp, deserializerFactory));
        }

        return list;
    }

    @Override public Class<?> modelClass() {
        throw new UnsupportedOperationException("The modelClass method shouldn't be used on ListDeserializer.");
    }
}

File: plugin/runtime/src/main/java/com/jaynewstrom/json/runtime/internal/ListDeserializer.kt
package com.jaynewstrom.json.runtime.internal

import com.fasterxml.jackson.core.JsonParser
import com.fasterxml.jackson.core.JsonToken
import com.jaynewstrom.json.runtime.JsonDeserializer
import com.jaynewstrom.json.runtime.JsonDeserializerFactory

import java.io.IOException

class ListDeserializer<T>(private val elementFactory: JsonDeserializer<T>) : JsonDeserializer<List<T>> {
    @Throws(IOException::class)
    override fun deserialize(jp: JsonParser, deserializerFactory: JsonDeserializerFactory): List<T> {
        if (jp.currentToken() != JsonToken.START_ARRAY) {
            throw IOException("Expected data to start with an Array")
        }

        val list = ArrayList<T>()

        while (jp.nextToken() != JsonToken.END_ARRAY) {
            list.add(elementFactory.deserialize(jp, deserializerFactory))
        }

        return list.toList()
    }
}
